### PR TITLE
 etcdutil, leadership: use RequestProgress in watch loop

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close https://github.com/tikv/pd/issues/6889
### What is changed and how does it work?

When connecting to multiple endpoints and injecting pd-leader-io-hang, the watch chan may get stuck, so I added a ticker to send `request progress` to ping whether the watch chan is working normally, and rebuild it if it doesn't receive messages from the watch chan for a long time.

And add `CheckWatchChan` to avoid block when creating a watch channel.

Use `logFields` to reduce code in log output

### Check List

Tests

- Unit test
- Manual test

run tpcc in pd-leader-io-hang case twenty times 
https://tcms.pingcap.net/dashboard/executions/plan/2130511
https://tcms.pingcap.net/dashboard/executions/plan/2100564

1. watch leader loop can exit normally
image_2023-08-22_11-04-15
image_2023-08-22_11-03-08

2. qps did not drop to zero and did not meet 9001 error

image
image

### Release note

```release-note
None.
```

### CODE REVIEW

**Comment by JmPotato** [comment]
Is this PR reviewable?

**Comment by lhy1024** [comment]
> Is this PR reviewable?

No, it still be need to more test

**Comment by lhy1024** [comment]
PTAL @JmPotato @rleungx

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

**Review by rleungx — APPROVED** [review]

File: pkg/election/leadership_test.go [hunk 1]
```diff
@@ -193,7 +268,8 @@ func checkExitWatch(t *testing.T, leaderKey string, injectFunc func(server *embe
 		done <- struct{}{}
 	}()
 
-	injectFunc(etcd, client2)
+	deferFunc := injectFunc(etcd, client2)
+	defer deferFunc()
```
[the remark this answers is not in the document]
Comment (lhy1024) [reply pkg/election/leadership_test.go:272]: They are different. We hope to run `injectFunc` first and run `deferFunc` to clean finally.

File: pkg/election/leadership.go [hunk 2]
```diff
@@ -256,15 +260,15 @@ func (ls *Leadership) Watch(serverCtx context.Context, revision int64) {
 			// When etcd is not available, the watcher.RequestProgress will block,
 			// so we check the etcd availability first.
 			if !etcdutil.IsHealthy(serverCtx, ls.client) {
-				log.Error("the connection of the leadership watcher is unhealthy",
+				log.Warn("the connection of the leadership watcher is unhealthy, retry to watch later",
```
Comment (rleungx) [inline pkg/election/leadership.go:272]: I think it's better to make the log different.

**Comment by lhy1024** [comment]
Each plan run 10 network-partition and 10 pd-leader-io-hang with 03e95db 
https://tcms.pingcap.net/dashboard/executions/plan/2101385
https://tcms.pingcap.net/dashboard/executions/plan/2101384

1. two follower exit leader loop normally
image
image
2. no qps drop and no 9001 error
image
image
3. no 9001 error in these two plans
image
image

**Comment by lhy1024** [comment]
test has updated, ptal @JmPotato @rleungx

File: pkg/election/leadership.go [hunk 3]
```diff
@@ -189,107 +192,139 @@ func (ls *Leadership) Watch(serverCtx context.Context, revision int64) {
...
+			log.Warn("the connection maybe unhealthy, retry to watch later", ls.watchLogFields(revision)...)
 			select {
 			case <-serverCtx.Done():
-				log.Info("server is closed, exit leader watch loop",
-					zap.String("leader-key", ls.leaderKey),
-					zap.String("purpose", ls.purpose))
+				log.Info("server is closed, exit leader watch loop", ls.logFields...)
 				return
 			case <-ticker.C:
-				// continue to check the etcd availability
```
Comment (rleungx) [inline pkg/election/leadership.go:237]: Why change it?

File: pkg/utils/etcdutil/etcdutil.go [hunk 4]
```diff
@@ -669,86 +674,135 @@ func (lw *LoopWatcher) initFromEtcd(ctx context.Context) int64 {
...
 		}
+		watcher = clientv3.NewWatcher(lw.client)
 		// In order to prevent a watch stream being stuck in a partitioned node,
 		// make sure to wrap context with "WithRequireLeader".
-		watchChanCtx, cancel := context.WithCancel(clientv3.WithRequireLeader(ctx))
-		watchChanCancel = cancel
-		opts := append(lw.opts, clientv3.WithRev(revision))
-		watchChan := watcher.Watch(watchChanCtx, lw.key, opts...)
-	WatchChanLoop:
+		watcherCtx, cancel := context.WithCancel(clientv3.WithRequireLeader(ctx))
```
Comment (rleungx) [inline pkg/utils/etcdutil/etcdutil.go:712]:
```suggestion
		watcherCtx, watchChanCancel = context.WithCancel(clientv3.WithRequireLeader(ctx))
```

**Review by JmPotato — APPROVED** [review]

File: pkg/utils/etcdutil/etcdutil.go [hunk 5]
```diff
@@ -840,3 +894,18 @@ func (lw *LoopWatcher) SetLoadTimeout(timeout time.Duration) {
 func (lw *LoopWatcher) SetLoadBatchSize(size int64) {
 	lw.loadBatchSize = size
 }
+
+// CheckWatchChan checks whether the watch channel is blocked for a long time while creating a new watch channel.
+func CheckWatchChan(ctx context.Context, cancel context.CancelFunc, done chan struct{}) {
```
Comment (rleungx) [inline pkg/utils/etcdutil/etcdutil.go:899]: What is the difference between checkStream?

File: pkg/utils/etcdutil/etcdutil.go (hunk 4 above)
Comment (lhy1024) [reply pkg/utils/etcdutil/etcdutil.go:712]: https://github.com/tikv/pd/pull/6577#discussion_r1267743604

File: pkg/utils/etcdutil/etcdutil.go (hunk 5 above)
Comment (lhy1024) [reply pkg/utils/etcdutil/etcdutil.go:899]: no difference? only avoid cycle import, how about moving `checkStream` out from `tsoutil`? Which path is suitable?

Comment (rleungx) [reply pkg/utils/etcdutil/etcdutil.go:899]: Maybe we can reuse it?

Comment (lhy1024) [reply pkg/utils/etcdutil/etcdutil.go]: move it to grpcutil

**Review by rleungx — APPROVED** [review]

**Pull request merged**